src/mlpack/methods/ann/layer/pixel_shuffle_impl.hpp

```diff
@@ -0,0 +1,152 @@
+/**
+ * @file methods/ann/layer/pixel_shuffle_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the PixelShuffle class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_PIXEL_SHUFFLE_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_PIXEL_SHUFFLE_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "pixel_shuffle.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PixelShuffle<InputDataType, OutputDataType>::PixelShuffle() :
+    upscaleFactor(0),
+    height(0),
+    width(0),
+    size(0),
+    batchSize(0),
+    outputHeight(0),
+    outputWidth(0),
+    sizeOut(0),
+    reset(false)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+PixelShuffle<InputDataType, OutputDataType>::PixelShuffle(
+    const size_t upscaleFactor,
+    const size_t height,
+    const size_t width,
+    const size_t size) :
+    upscaleFactor(upscaleFactor),
+    height(height),
+    width(width),
+    size(size),
+    batchSize(0),
+    outputHeight(0),
+    outputWidth(0),
+    sizeOut(0),
+    reset(false)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void PixelShuffle<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  if (!reset)
+  {
+    batchSize = input.n_cols;
+    sizeOut = size / std::pow(upscaleFactor, 2);
+    outputHeight = height * upscaleFactor;
+    outputWidth = width * upscaleFactor;
+    reset = true;
+  }
+  output.zeros(outputHeight * outputWidth * sizeOut, batchSize);
+  for (size_t n = 0; n < batchSize; n++)
+  {
+    arma::mat inputImage = input.col(n);
+    arma::mat outputImage = output.col(n);
+    arma::cube inputTemp(const_cast<arma::mat&>(inputImage).memptr(), height,
+        width, size, false, false);
+    arma::cube outputTemp(const_cast<arma::mat&>(outputImage).memptr(),
+        outputHeight, outputWidth, sizeOut, false, false);
```

Comment:
I'm not sure about this but maybe we can do this without the creating copies (which add to latency / space utilization). If 
```
arma::cube inputTemp(const_cast<arma::mat&>(input.col(n)).memptr(), height,
          width, size, false, false);
```
doesn't work, we can 
```
arma::cube inputTemp(const_cast<arma::mat&>(input).memptr(), height,
          width, size * batchSize, false, false);
```
and then use span. That way we don't create copies.
Also save time on copying back (maybe). Let me know incase I missed something.

Reply:
@kartikdutt18 @zoq I made the required changes.  Now the copying step has been removed. Please  review.